### ISSUE TITLE
OrderBy on Aggregation without Group By

### ISSUE THREAD

**Issue by jakesyl** [ClickHouse/ClickHouse#11459]
**tl;dr: Is there any way to do an order by on a materialized view without a group by**

We've got a few billion transactions of various values we sum to get a live `balance` for a given `users` in a materialized view. We then `ORDER BY` this `balance` in a materialized view and allow pagination. `GROUP BY`  is too slow so we'd like to pre-compute the data somehow.

## The Schema
We have a table that looks something like this:

```sql
CREATE TABLE IF NOT EXISTS transactions
	(
                 txid FixedString(32),
		 value Float64,
		 user String
	) ENGINE = AggregatingMergeTree()
	ORDER BY (user, txid, value)
	PRIMARY KEY (user, txid, value)
```

There are about 4 billion rows in here and we're adding about 7 million per day. We have a materialized view off of this table (using [`TO`](https://www.altinity.com/blog/clickhouse-materialized-views-illuminated-part-2)) that keeps track of transaction count, balance, and transaction count of these users. I've tried out a bunch of different implementations for this ([`MergeTree`](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/mergetree/) [`SummingMergeTree`](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/summingmergetree/), [`AggregatingMergeTree`](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/aggregatingmergetree/), [`ReplacingMergeTree`](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/replacingmergetree/)) but below,  I'll use the simplest iteration of the table:

```sql
CREATE TABLE IF NOT EXISTS users (
		balance AggregateFunction(sum, Float64),
		user String,
		tx_count AggregateFunction(uniq, FixedString(64)),
	)
	ENGINE = AggregatingMergeTree()
	ORDER BY (user, balance);
```
I do an initial insert like this:

```sql
INSERT INTO users
SELECT
		sum(value) as balance,
		user,
                COUNT(*) as tx_count
FROM transactions
GROUP BY user
```

```sql
CREATE MATERIALIZED VIEW users
	TO users
	AS SELECT
		sum(value) as balance,
		user,
               COUNT(*) as tx_count
FROM transactions
GROUP BY user -- I do have a where here to make sure this is only new data, 
             -- schema is simplified for github
```

I then have a materialized view using [`TO`](https://www.altinity.com/blog/clickhouse-materialized-views-illuminated-part-2) to populate this table. No matter what I do here I end up with duplicate records. All I'd like here is to be able to query ~700 million rows using and `ORDER BY balance` in a performant manner. As soon as I add in `GROUP BY` this slows to a halt.

I realize above might not be correct (I'm not even sure if `SimpleAggregateFunction`) is meant to deduplicate here. It's just the simplest implementation to demonstrate the problem. Here are some of the solutions I've tried:

## What I've Tried so Far

- [`SummingMergeTree`](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/summingmergetree/):   will allow me to deduplicate in the background (and this works), but still requires the `GROUP BY` query which is too slow ~20 seconds and took up too much memory. 
- [`AggregatingMergeTree`](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/aggregatingmergetree/) with [sumState](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/aggregatingmergetree/#example-of-an-aggregated-materialized-view): As per #10326, tried using a sumState. This still required a group by and was less performant than `SummingMergeTree` (presumably because of the merging)
- [`ReplacingMergeTree`](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/replacingmergetree/) I tried the same thing as `SummingMergeTree` and this was predictably less performant 
- [`LiveView`](https://github.com/ClickHouse/ClickHouse/pull/3925): [Creating a live view](https://www.altinity.com/blog/2019/11/13/making-data-come-to-life-with-clickhouse-live-view-tables) on the select query above. This is not supported as per #3925
- Query time deduplication: In this implementation, I tried to query from a `subquery` which eliminated the duplicates in a group by with no luck #10876
- Query time max: In this implementation, I introduced a `CREATED_BY` and tried a `MAX()` on this field with no performance improvements
- `OPTIMIZE DEDUPLICATE`/ `OPTIMIZE DEDUPLICATE FINAL`- I didn't expect this to work, and it didn't as per #1178, #5125, etc.
` SELECT FINAL` - less performant than group by everywhere.

## What I'm going to try
So I think I understand the fundamental problems here:

- Clickhouse is append only (deletes/updates are expensive)
- MVs are therefore triggers (and have no context on underlying data)
- Triggers are appending new data and old data can't be deleted synchronously.

Also keep in mind my data pipeline is not streaming - we're talking about an update every ~15 seconds. Here are my ideas. I'll be updating this issue as I try things:

1. Query time validation: e.g. `SELECT * FROM x ORDER BY BALANCER FROM (SELECT * FROM x ORDER BY BALANCE DESC  INNER JOIN ON address argMax(balance) as balancer LIMIT 50))`
1. [Skipping Index](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/mergetree/#table_engine-mergetree-data_skipping-indexes) on tx_count with `max(tx_count)` on GROUP BY with a `ReplicatedReplacingMergeTree` and version equal to `tx_count` (with [`argMax`](https://stackoverflow.com/a/59788716/1011803))
1. ~Speed up group by using [`LowCardinality`](https://github.com/ClickHouse/ClickHouse/issues/4796#issuecomment-477610482) as per #4796/#3138~ 90 seconds instead of 140, nice improvement
1. ~Lower Index Granularity (currently at 8192) - trying out 50 combined with `LowCardinality`~ slows it down by ~40 seconds. Makes sense as per [this explanation](https://medium.com/@f1yegor/clickhouse-primary-keys-2cf2a45d7324)
1. ~Speed up group by using `FixedString` on `User` (this will average appxm. 32 null bytes)~ pretty good results here. 250 seconds vs 139 seconds
1. ~`FixedString`/`LowCardinality` combo~ - OOM because of the `null_bytes`
1. ~Remove replication~: all of our tables are `Replicated`, but none actually replicate so as expected this didn't do anything 
1. ~Only sum on column I want to paginate on~ (60 seconds instead of 500 so definitely an improvement. Might work with `LowCardinality` and ~`FixedString`~, will test there as well
1. Full reinsertion [CollapsingMergeTree](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/versionedcollapsingmergetree/). Every interval reinserts all previously computed data (using group by on MV) with a version NOW() column.  WHERE time=NOW() is used as a filter for our oderby and users are collapsed according to `NOW()` since primary key is `User`
1. ~Create a MV to insert into a Memory table on every update (since these can be treated as temporary)~ Too big
1. ~Two layer mv: `maxState on summingMergeTree`~ same duplicate issue
1.  Create/Drop Partition: Recompute _all_ users every block, drop old partitions, create a new one
1. ~Live View (I'm not sure how to pull this off)~ - won't work as per my [comment](https://github.com/ClickHouse/ClickHouse/pull/5541#issuecomment-639194808) on #5541 
1. Historical balance table with some sort of cumulative sum (so all users balances get recomputed at every interval)~ ([VersionedCollapsingMergeTree](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/versionedcollapsingmergetree/) works just as well here and we have [deduplicate_blocks_in_dependent_materialized_views](https://clickhouse.tech/docs/en/operations/settings/settings/#settings-insert-deduplicate) and [insert_deduplicate](https://clickhouse.tech/docs/en/operations/settings/settings/#settings-insert-deduplicate) on so these rows should be inserted in a single transaction. The examples precedent I've seen for this use (#4628 and [percona](https://www.percona.com/sites/default/files/ple19-slides/day2-am/clickhouse-materialized-views-a-secret-weapon-for-high-performance-analytics.pdf))
1. ~Calculate position at insert time (using group by)~ window views aren't possible at insert time and have the same issues as other materialized views
1. [VictoriaMetrics style Snapshots](https://medium.com/@valyala/how-victoriametrics-makes-instant-snapshots-for-multi-terabyte-time-series-data-e1f3fb0e0282) - (hard links, these are super cool by the way)
1. Merge in #9113 (Optimization of GROUP BY with respect to table sorting key) locally and see if group by is fast enough
1. Create a custom non-LSM tree engine (or maybe mysql) that supports updates since aggregated data doesn't change that often
1.  New MV's every time at interval (physically `recreate MATERIALIZED VIEW` with `POPULATE`)
1. Store the data somewhere else: (e.g. redis, mysql) w/ update performance (*last resort*)

Would appreciate it if someone could point me in the right direction here

**Comment by den-crane** [comment]
If you need something like 'select ... order by balance desc' with millisecond realtime response in streaming manner the only way is to recalculate it regularly (e.g. using cron) and store a ready result in another table. There is no Engine which can solve it and I don't see how it can be implemented with MergeTree architecture.

**Comment by jakesyl** [comment]
Got it @den-crane is there some sort of way to store the ready result so it's queryable or should I just be dropping/recreating tables? (Asking as I know the fundamental problem here is deletes are expensive in LSM(

**Comment by vzakaznikov** [comment]
I will look into adding proper support for `PREWHERE` in LiveView tables and should have a PR ready by the end of next week if there are no major road blocks.

**Comment by den-crane** [comment]
@jakesyl you can insert into an intermediate table,  drop a partition in the main table, and attach / move from the intermediate. Or you can rename tables (do swap in a single rename).

**Comment by jakesyl** [comment]
@den-crane appreciate the reccomendation. Trying this right now and having some issues around write performance but will let you know how this goes

**Comment by jakesyl** [comment]
I have one other idea here. What about something like

```sql
SELECT
	argMax(balance, tx_count) as balance,
	any(user) as user,
	max(tx_count) as tx_count
FROM users -- or subquery
GROUP BY user
ORDER BY balance
```

**Comment by jakesyl** [comment]
So I applied some of the optimizations above (namely [`LowCardinality`](https://github.com/ClickHouse/ClickHouse/issues/4796#issuecomment-477610482), [`CollapsingMergeTree`](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/versionedcollapsingmergetree/). (with `tx_count` as version), [skipping indices](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/mergetree/#table_engine-mergetree-data_skipping-indexes) (on `tx_count`), and will be following @den-crane's recommendation regarding cron inserts (pending `LiveView`). 

I've got one other question. Right now I'm doing

```
SELECT
    max(tx_count),
    argMax(balance, tx_count) AS balance,
    user
    -- I have a 4 other columns in here that all require argMax()
FROM users
GROUP BY address
```

and inserting into my partitioned pre-computed table. Is there a faster way to do this so I can avoid `argMax` a bunch of times or does this get optimized out automatically?

**Comment by vzakaznikov** [comment]
> @vzakaznikov does the above seem like it would work with a `LiveView` (considering number of rows?)
@jakesyl, I am not sure. You will have to give it a try. From the first glance it seems like what you need is MV. Live views do not store any data and just keep partial query result in memory. So it really depends on how big is your query result and if you can recompute the initial result fast enough when the server restarts.  Also, given that you have GROUP BY it might not work as expected. I don't remember the details about GROUP BY with live views and will need to write a test case for it to see how it behaves and if there are any problems or workarounds using sub-queries.

**Comment by jakesyl** [comment]
Yeah - we might just need to throw some more ram at this either way. `LiveView` is an abstraction over @den-crane's solution since they're both storing results in memory for live querying. Appreciate the response

**Comment by vzakaznikov** [comment]
> Yeah - we might just need to throw some more ram at this either way. `LiveView` is an abstraction over @den-crane's solution since they're both storing results in memory for live querying. Appreciate the response

@jakesyl, if you did want to try to get live views to work with your case, we would first need to create a simple scenario that mimics what you are trying to do. For an example, you can take a look at https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/00973_live_view_select.sql. Once we have an exact minimal scenario then we can easily work with it.

**Comment by PyCht** [comment]
> If you need something like 'select ... order by balance desc' with millisecond realtime response in streaming manner the only way is to recalculate it regularly (e.g. using cron) and store a ready result in another table. There is no Engine which can solve it and I don't see how it can be implemented with MergeTree architecture.

@den-crane would window view work well in this scenario? (replacing cron)
